.github/workflows/podvm_mkosi.yaml

```diff
@@ -19,81 +46,140 @@ defaults:
     working-directory: src/cloud-api-adaptor
 
 jobs:
-  build-binaries:
-    name: Build binaries
-    if : ${{ inputs.binaries-image == '' }}
-    runs-on: ubuntu-24.04
+  build-image:
+    name: Build mkosi image
+    runs-on: ${{ inputs.arch == 's390x' && 's390x' || 'ubuntu-24.04' }}
     permissions:
       contents: read
       packages: write
     steps:
       - name: Checkout
         uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+          ref: "${{ inputs.git_ref }}"
+
+      - name: Rebase the code
+        if: github.event_name == 'pull_request_target'
+        working-directory: ./
+        run: |
+          ./hack/ci-helper.sh rebase-atop-of-the-latest-target-branch
 
       - name: Set up Docker Buildx
         uses: docker/setup-buildx-action@v3
 
-      - name: Log in to GitHub Container Registry
+      - name: Login to quay Container Registry
+        if: ${{ startsWith(inputs.registry, 'quay.io') }}
+        uses: docker/login-action@v3
+        with:
+          registry: quay.io
+          username: ${{ secrets.QUAY_USERNAME }}
+          password: ${{ secrets.QUAY_PASSWORD }}
+
+      - name: Login to the ghcr Container registry
+        if: ${{ startsWith(inputs.registry, 'ghcr.io') }}
         uses: docker/login-action@v3
         with:
           registry: ghcr.io
           username: ${{ github.actor }}
           password: ${{ secrets.GITHUB_TOKEN }}
 
-      - name: Build builder
-        uses: docker/build-push-action@v6
+      - name: Install build dependencies
+        run: |
+          sudo apt-get update -y
+          sudo apt-get install -y bubblewrap alien dnf qemu-utils uidmap
+          sudo snap install yq
+
+      - name: Read properties from versions.yaml
+        if: ${{ inputs.arch == 's390x' }}
+        run: |
+          go_version="$(yq '.tools.golang' versions.yaml)"
+          echo "GO_VERSION=${go_version}" >> "$GITHUB_ENV"
+
+      - name: Setup Golang version ${{ env.GO_VERSION }}
+        if: ${{ inputs.arch == 's390x' }}
+        uses: actions/setup-go@v5
         with:
-          tags: ghcr.io/${{ github.repository }}/podvm/builder-fedora:${{ github.sha }}
-          context: src
-          file: src/cloud-api-adaptor/podvm/Dockerfile.podvm_builder.fedora
-          push: true
+          go-version: ${{ env.GO_VERSION }}
 
-      # Build binaries need git commit id as part of image name
-      - name: Prepare .git folder
-        working-directory: ./
+      - name: Build builder
+        id: build_builder
+        working-directory: src/cloud-api-adaptor/podvm-mkosi
         run: |
-          cp -rf .git src/.git
+          # If the builder image matching our git sha exists, then skip this build
+          builder_image=${{ inputs.registry }}/podvm-builder-fedora-${{ inputs.arch }}
+          tag=${{ github.sha }}
```

Comment:
Hi @stevenhorsman !

I'd like to review the use of `github.sha` throughout the workflow. Default case `actions/checkout@v4` checks out the code from `git_ref` (mandatory workflow input). Does  `actions/checkout@v4` update the value of `github.sha`? (I think so, but double-checking...). Even if `github.sha` is updated by `actions/checkout@v4`; there is the case of `pull_request_target` where it `rebase-atop-of-the-latest-target-branch`, is `github.sha` correct?

Reply:
Nice points - do you think that something like `git rev-parse --short HEAD` would be better? Part of me worries about the sha not clearly mapping back to a test commit, which might back the transparency of code slightly tougher, but that shouldn't be a problem for non-PR branches where it's more important and I think having the updated builder image is more important, so it's a worthwhile change.

Reply:
Ok, I think I've address this now